File: day_03/day_03.py
import re
import aocutils

@aocutils.timeFunction
def main():
    total1 = 0
    total2 = 0
    do = 1

    lines = aocutils.readFile(3, False)
    for line in lines:
        muls = re.findall(r"mul\((\d+),(\d+)\)|(do\(\))|(don't\(\))", line)
        for match in muls:
            # part 1 and 2
            if match[0]:
                total1 += int(match[0]) * int(match[1])
                total2 += do * int(match[0]) * int(match[1])
            
            # part 2
            elif match[2]: do = 1
            else: do = 0
    
    aocutils.printParts(total1, total2)

if __name__ == "__main__":
    main()

File: day_03/day_03.go
package day_03

import (
	a "aoc_2024/aocutils"
	"fmt"
	"regexp"
)

func Day_03() {
	total1, total2, do := 0, 0, 1
	lines := a.Reader(3)
	reg := regexp.MustCompile(`mul\((\d+),(\d+)\)|(do\(\))|(don't\(\))`)
	for _, line := range lines {
		matches := reg.FindAllStringSubmatch(line, -1)
		for _, match := range matches {
			if match[0] == "don't()" {
				do = 0
			} else if match[0] == "do()" {
				do = 1
			} else {
				total1 += a.ToI(match[1]) * a.ToI(match[2])
				total2 += do * a.ToI(match[1]) * a.ToI(match[2])
			}
		}
	}

	fmt.Printf("Part 1: %d\n", total1)
	fmt.Printf("Part 2: %d\n", total2)
}